teams/vsa/engineering/forms-system/custom_component_multiple.md

```diff
@@ -0,0 +1,138 @@
+});
+const mapDispatchToProps = {
+  setData,
+};
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps,
+)(ReviewCardField);
+
+  ```
+
+## Side-Effect: Extra Component
+You still need to pass in a schema that accounts for the data you want to modify with the form system.  The form system will then add its own component for that data.  So you end up with an extra component.  We need to hide it!  The original way to do this was to leverage a hideIf param that exits in the uiSchema.  There is, however, a gotcha with this.  When you hide a component on the form, it removes its schema thus wiping out its form state.
+
+*[src/platform/forms-system/src/js/state/helpers.js](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/state/helpers.js)*
+```
+export function expandArrayPages(pageList, data) {
+. . .
+     } else if (nextPage.arrayPath !== lastArrayPath && !!arrayPages.length) {
+        const newList = currentList.concat(
+          generateArrayPages(arrayPages, data),
+          nextPage,
+        );
+ ```
+
+## Workaround: Hide the Extra Component 
+Instead of using a hideIf func, you can pass a custom className to the uiSchema.  So I passed in the utility className from the design system to just hide it via CSS. \
+*Discussion on slack: https://dsva.slack.com/archives/CBU0KDSB1/p1587597608400600*
+
+## Limitations
+When we pre-fill data into a field that's hidden by default, that data gets removed before the user gets to see it. It's just not even there. We were unable to use selectable cards due to this limitation with expanding check boxes. By design, hidden data gets removed in *recalculateSchemaAndData* because... well, it's hidden. The design decision behind this is so we don't send any information downstream that the veteran didn't get a chance to review or sort of "opted out" of that branch. We haven't run into this before because we've never pre-filled data into a field hidden by default because it goes against the design pattern for all the forms previously. Workaround to this limitation: rather than hiding, make a new page where the veteran can select the items, depending on the answer to that question. This is the pattern the 526 all claims uses for selecting rated disabilities. \
+*Discussion on slack: https://dsva.slack.com/archives/CBU0KDSB1/p1588080662115800*
+
+## Why It's a Hack
+If it isn't obvious by now, rather than treating VSFS/RJSF as a back box, we're having to peek inside the form's internal Redux state (gray box). Since this wasn't thought through by design, you'll run into limitations such as the hidden field scenario we just mentioned.
+
+## Background: The Original Problem
+Here is the original problem we were trying to solve, and conversations we had in Slack related to it. First, here is the write-up from [mr0sari0](https://github.com/mr0sari0):
+
+### Form System Questions/Limitations
+Sharing state between `ui:fields`,
```

Comment:
@bradl3yC and @mr0sari0 is this accurate? Are the ui:fields under the same parent, or do they belong to or share state under different react components?

Reply:
This not - This should be done using `ui:widgets`

Reply:
ah, so whenever we originally said "ui:fields" we should have communicated something like, "ui:fields in different ui:widgets => ui:widgets" b/c only saying "ui:fields" would imply fields within the same widget?

Reply:
@mr0sari0 knowing this, should we change:

Sharing state between `ui:fields`,
- A field can not trigger a change to another field
- Two schema fields cannot alter the state of each other within a `ui:field`.

to be this:

Sharing state between `ui:widgets`,
- A field can not trigger a change to another field that is in a different widget
- Two schema fields cannot alter the state of each other across widgets

?

Reply:
I was saying it should be `ui:widgets` because thats the key that custom components should be used under.  It works with a ui:field but its not how you should use that particular key/value.

Reply:
If widgets are preferred as best practice,I can update the doc to be widgets instead of fields.